{"cells":[{"cell_type":"markdown","metadata":{"id":"G4yBrceuFbf3"},"source":["<img src=\"https://raw.githubusercontent.com/sokrypton/ColabFold/main/.github/ColabFold_Marv_Logo_Small.png\" height=\"200\" align=\"right\" style=\"height:240px\">\n","\n","##ColabFold v1.5.2-patch: AlphaFold2 using MMseqs2\n","\n","Easy to use protein structure and complex prediction using [AlphaFold2](https://www.nature.com/articles/s41586-021-03819-2) and [Alphafold2-multimer](https://www.biorxiv.org/content/10.1101/2021.10.04.463034v1). Sequence alignments/templates are generated through [MMseqs2](mmseqs.com) and [HHsearch](https://github.com/soedinglab/hh-suite). For more details, see <a href=\"#Instructions\">bottom</a> of the notebook, checkout the [ColabFold GitHub](https://github.com/sokrypton/ColabFold) and read our manuscript.\n","Old versions: [v1.4](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.4.0/AlphaFold2.ipynb), [v1.5.1](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.5.1/AlphaFold2.ipynb)\n","\n","[Mirdita M, Schütze K, Moriwaki Y, Heo L, Ovchinnikov S, Steinegger M. ColabFold: Making protein folding accessible to all.\n","*Nature Methods*, 2022](https://www.nature.com/articles/s41592-022-01488-1)\n","\n","-----------\n","\n","### News\n","- <b><font color='green'>2023/07/31: The ColabFold MSA server is back to normal. It was using older DB (UniRef30 2202/PDB70 220313) from 27th ~8:30 AM CEST to 31st ~11:10 AM CEST.</font></b>\n","- <b><font color='green'>2023/06/12: New databases! UniRef30 updated to 2023_02 and PDB to 230517. We now use PDB100 instead of PDB70 (see [notes](#pdb100)).</font></b>\n","- <b><font color='green'>2023/06/12: We introduced a new default pairing strategy: Previously, for multimer predictions with more than 2 chains, we only pair if all sequences taxonomically match (\"complete\" pairing). The new default \"greedy\" strategy pairs any taxonomically matching subsets.</font></b>"]},{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":92869,"status":"ok","timestamp":1694805860975,"user":{"displayName":"Itzayana Chavarría Quintanilla","userId":"15426686823544663593"},"user_tz":300},"id":"AzIKiDiCaHAn","outputId":"b1b0200d-f675-45e9-f03d-d83142013650"},"outputs":[{"output_type":"stream","name":"stdout","text":["installing colabfold...\n","installing conda...\n","installing hhsuite and amber...\n","CPU times: user 327 ms, sys: 60.7 ms, total: 388 ms\n","Wall time: 1min 32s\n"]}],"source":["#@title Install dependencies\n","%%time\n","import os\n","from sys import version_info\n","python_version = f\"{version_info.major}.{version_info.minor}\"\n","\n","USE_AMBER = True\n","USE_TEMPLATES = True\n","PYTHON_VERSION = python_version\n","\n","if not os.path.isfile(\"COLABFOLD_READY\"):\n","  print(\"installing colabfold...\")\n","  os.system(\"pip install -q --no-warn-conflicts 'colabfold[alphafold-minus-jax] @ git+https://github.com/sokrypton/ColabFold'\")\n","  os.system(\"pip install --upgrade dm-haiku\")\n","  os.system(\"ln -s /usr/local/lib/python3.*/dist-packages/colabfold colabfold\")\n","  os.system(\"ln -s /usr/local/lib/python3.*/dist-packages/alphafold alphafold\")\n","  # patch for jax > 0.3.25\n","  os.system(\"sed -i 's/weights = jax.nn.softmax(logits)/logits=jnp.clip(logits,-1e8,1e8);weights=jax.nn.softmax(logits)/g' alphafold/model/modules.py\")\n","  os.system(\"touch COLABFOLD_READY\")\n","\n","if USE_AMBER or USE_TEMPLATES:\n","  if not os.path.isfile(\"CONDA_READY\"):\n","    print(\"installing conda...\")\n","    os.system(\"wget -qnc https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh\")\n","    os.system(\"bash Mambaforge-Linux-x86_64.sh -bfp /usr/local\")\n","    os.system(\"mamba config --set auto_update_conda false\")\n","    os.system(\"touch CONDA_READY\")\n","\n","if USE_TEMPLATES and not os.path.isfile(\"HH_READY\") and USE_AMBER and not os.path.isfile(\"AMBER_READY\"):\n","  print(\"installing hhsuite and amber...\")\n","  os.system(f\"mamba install -y -c conda-forge -c bioconda kalign2=2.04 hhsuite=3.3.0 openmm=7.7.0 python='{PYTHON_VERSION}' pdbfixer\")\n","  os.system(\"touch HH_READY\")\n","  os.system(\"touch AMBER_READY\")\n","else:\n","  if USE_TEMPLATES and not os.path.isfile(\"HH_READY\"):\n","    print(\"installing hhsuite...\")\n","    os.system(f\"mamba install -y -c conda-forge -c bioconda kalign2=2.04 hhsuite=3.3.0 python='{PYTHON_VERSION}'\")\n","    os.system(\"touch HH_READY\")\n","  if USE_AMBER and not os.path.isfile(\"AMBER_READY\"):\n","    print(\"installing amber...\")\n","    os.system(f\"mamba install -y -c conda-forge openmm=7.7.0 python='{PYTHON_VERSION}' pdbfixer\")\n","    os.system(\"touch AMBER_READY\")"]},{"cell_type":"code","execution_count":2,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":23709,"status":"ok","timestamp":1694805884668,"user":{"displayName":"Itzayana Chavarría Quintanilla","userId":"15426686823544663593"},"user_tz":300},"id":"We6xHnVxNuLp","outputId":"e426e6db-14e6-4c6a-8328-c607729cc135"},"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n","You are logged into Google Drive and are good to go!\n"]}],"source":["#Imports and default config\n","\n","from google.colab import drive\n","from google.colab import files\n","import os\n","import re\n","import hashlib\n","import random\n","\n","from pydrive.drive import GoogleDrive\n","from pydrive.auth import GoogleAuth\n","from google.colab import auth\n","from oauth2client.client import GoogleCredentials\n","import sys\n","import warnings\n","warnings.simplefilter(action='ignore', category=FutureWarning)\n","from Bio import BiopythonDeprecationWarning\n","warnings.simplefilter(action='ignore', category=BiopythonDeprecationWarning)\n","from pathlib import Path\n","from colabfold.download import download_alphafold_params, default_data_dir\n","from colabfold.utils import setup_logging\n","from colabfold.batch import get_queries, run, set_model_type\n","from colabfold.plot import plot_msa_v2\n","\n","import os\n","import numpy as np\n","try:\n","  K80_chk = os.popen('nvidia-smi | grep \"Tesla K80\" | wc -l').read()\n","except:\n","  K80_chk = \"0\"\n","  pass\n","if \"1\" in K80_chk:\n","  print(\"WARNING: found GPU Tesla K80: limited to total length < 1000\")\n","  if \"TF_FORCE_UNIFIED_MEMORY\" in os.environ:\n","    del os.environ[\"TF_FORCE_UNIFIED_MEMORY\"]\n","  if \"XLA_PYTHON_CLIENT_MEM_FRACTION\" in os.environ:\n","    del os.environ[\"XLA_PYTHON_CLIENT_MEM_FRACTION\"]\n","\n","from colabfold.colabfold import plot_protein\n","from pathlib import Path\n","import matplotlib.pyplot as plt\n","\n","# For some reason we need that to get pdbfixer to import\n","if USE_AMBER and f\"/usr/local/lib/python{python_version}/site-packages/\" not in sys.path:\n","    sys.path.insert(0, f\"/usr/local/lib/python{python_version}/site-packages/\")\n","\n","#Login to drive\n","drive.mount('/content/drive')\n","print(\"You are logged into Google Drive and are good to go!\")"]},{"cell_type":"code","execution_count":3,"metadata":{"executionInfo":{"elapsed":18,"status":"ok","timestamp":1694805884670,"user":{"displayName":"Itzayana Chavarría Quintanilla","userId":"15426686823544663593"},"user_tz":300},"id":"2wp80Yj_VQg5"},"outputs":[],"source":["# Methods and more\n","def add_hash(x,y):\n","  return x+\"_\"+hashlib.sha1(y.encode()).hexdigest()[:5]\n","\n","# check if directory with jobname exists\n","def check(folder):\n","  if os.path.exists(folder):\n","    return False\n","  else:\n","    return True\n","\n","def set_basic_config():\n","  num_relax = 0 #param [0, 1, 5] {type:\"raw\"}\n","  template_mode = \"none\" #param [\"none\", \"pdb100\",\"custom\"]\n","  _use_amber = num_relax > 0\n","  num_seeds = 1 #param [1,2,4,8,16] {type:\"raw\"}\n","  use_dropout = False #param {type:\"boolean\"}\n","  num_models = 1\n","\n","\n","  #markdown #### Save settings\n","  save_all = False #param {type:\"boolean\"}\n","  save_recycles = False #param {type:\"boolean\"}\n","  #markdown -  if the save_to_google_drive option was selected, the result zip will be uploaded to your Google Drive\n","\n","  if template_mode == \"pdb100\":\n","    _use_templates = True\n","    custom_template_path = None\n","  else:\n","    custom_template_path = None\n","    _use_templates = False\n","\n","  msa_mode = \"mmseqs2_uniref_env\" #param [\"mmseqs2_uniref_env\", \"mmseqs2_uniref\",\"single_sequence\",\"custom\"]\n","  pair_mode = \"unpaired_paired\" #param [\"unpaired_paired\",\"paired\",\"unpaired\"] {type:\"string\"}\n","\n","  # # decide which a3m to use\n","  # if \"mmseqs2\" in msa_mode:\n","  #   a3m_file = os.path.join(jobname,f\"{jobname}.a3m\")\n","\n","  # else:\n","  #   a3m_file = os.path.join(jobname,f\"{jobname}.single_sequence.a3m\")\n","  #   with open(a3m_file, \"w\") as text_file:\n","  #     text_file.write(\">1\\n%s\" % query_sequence)\n","\n","  #markdown ### Advanced settings\n","  model_type = \"auto\" #param [\"auto\", \"alphafold2_ptm\", \"alphafold2_multimer_v1\", \"alphafold2_multimer_v2\", \"alphafold2_multimer_v3\"]\n","  #markdown - if `auto` selected, will use `alphafold2_ptm` for monomer prediction and `alphafold2_multimer_v3` for complex prediction.\n","  #markdown Any of the mode_types can be used (regardless if input is monomer or complex).\n","  num_recycles = \"auto\" #param [\"auto\", \"0\", \"1\", \"3\", \"6\", \"12\", \"24\", \"48\"]\n","  num_recycles = None if num_recycles == \"auto\" else int(num_recycles)\n","  recycle_early_stop_tolerance = \"auto\" #param [\"auto\", \"0.0\", \"0.5\", \"1.0\"]\n","  recycle_early_stop_tolerance = None if recycle_early_stop_tolerance == \"auto\" else float(recycle_early_stop_tolerance)\n","  #markdown - if `auto` selected, will use 20 recycles if `model_type=alphafold2_multimer_v3` (with tol=0.5), all else 3 recycles (with tol=0.0).\n","  pairing_strategy = \"greedy\" #param [\"greedy\", \"complete\"] {type:\"string\"}\n","  #markdown - `greedy` = pair any taxonomically matching subsets, `complete` = all sequences have to match in one line.\n","\n","  max_msa = \"auto\" #param [\"auto\", \"512:1024\", \"256:512\", \"64:128\", \"32:64\", \"16:32\"]\n","  if max_msa == \"auto\": max_msa = None\n","\n","  return dict(\n","      max_msa = max_msa,\n","      use_templates = _use_templates,\n","      custom_template_path = custom_template_path,\n","      num_relax = num_relax,\n","      msa_mode = msa_mode,\n","      num_recycles = num_recycles,\n","      recycle_early_stop_tolerance = recycle_early_stop_tolerance,\n","      num_seeds = num_seeds,\n","      use_dropout = use_dropout,\n","      pair_mode = pair_mode,\n","      pairing_strategy = pairing_strategy,\n","      save_all = save_all,\n","      save_recycles = save_recycles,\n","      model_type = model_type,\n","      num_models = num_models\n","      )"]},{"cell_type":"code","execution_count":4,"metadata":{"executionInfo":{"elapsed":16,"status":"ok","timestamp":1694805884670,"user":{"displayName":"Itzayana Chavarría Quintanilla","userId":"15426686823544663593"},"user_tz":300},"id":"Kq56ssJVRilv"},"outputs":[],"source":["def get_config(fasta_file_path: str, results_dir: str):\n","  queries, is_complex = get_queries(fasta_file_path)\n","  jobname = os.path.basename(fasta_file_path).split(\".\")[0]\n","  result_dir = os.path.join(results_dir, jobname)\n","  config = dict(\n","      queries = queries,\n","      is_complex = is_complex,\n","      result_dir = result_dir\n","      )\n","  config.update(set_basic_config())\n","  return config"]},{"cell_type":"code","execution_count":5,"metadata":{"executionInfo":{"elapsed":17,"status":"ok","timestamp":1694805884672,"user":{"displayName":"Itzayana Chavarría Quintanilla","userId":"15426686823544663593"},"user_tz":300},"id":"mbaIO9pWjaN0"},"outputs":[],"source":["\n","def run_job(config):\n","  result_dir = config[\"result_dir\"]\n","  # log_filename = os.path.join(result_dir,\"log.txt\")\n","  # if not os.path.isfile(log_filename) or 'logging_setup' not in globals():\n","  #   setup_logging(Path(log_filename))\n","  #   logging_setup = True\n","\n","\n","  model_type = set_model_type(config[\"is_complex\"], config[\"model_type\"])\n","\n","  if \"multimer\" in model_type and config[\"max_msa\"] is not None:\n","    use_cluster_profile = False\n","  else:\n","    use_cluster_profile = True\n","\n","  download_alphafold_params(model_type, Path(\".\"))\n","  results = run(\n","      queries=config[\"queries\"],\n","      result_dir=result_dir,\n","      use_templates=config[\"use_templates\"],\n","      custom_template_path=config[\"custom_template_path\"],\n","      num_relax=config[\"num_relax\"],\n","      msa_mode=config[\"msa_mode\"],\n","      model_type=model_type,\n","      num_models=config[\"num_models\"],\n","      num_recycles=config[\"num_recycles\"],\n","      recycle_early_stop_tolerance=config[\"recycle_early_stop_tolerance\"],\n","      num_seeds=config[\"num_seeds\"],\n","      use_dropout=config[\"use_dropout\"],\n","      model_order=[1,2,3,4,5],\n","      is_complex=config[\"is_complex\"],\n","      data_dir=Path(\".\"),\n","      keep_existing_results=False,\n","      rank_by=\"auto\",\n","      pair_mode=config[\"pair_mode\"],\n","      pairing_strategy=config[\"pairing_strategy\"],\n","      stop_at_score=float(100),\n","      # prediction_callback=prediction_callback,\n","      zip_results=False,\n","      save_all=config[\"save_all\"],\n","      max_msa=config[\"max_msa\"],\n","      use_cluster_profile=use_cluster_profile,\n","      # input_features_callback=input_features_callback,\n","      save_recycles=config[\"save_recycles\"],\n","  )"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"CfqshuILmd6Q","outputId":"8a80c40a-4147-41f1-8672-c24f3e4162c1"},"outputs":[{"output_type":"stream","name":"stdout","text":["COMPUTING RESULTS FOR euskadiensis_alpha16mannotransfeoch1like.fasta\n"]},{"output_type":"stream","name":"stderr","text":["Downloading alphafold2 weights to .:  36%|███▌      | 1.25G/3.47G [00:34<00:58, 41.1MB/s]"]}],"source":["from typing import List\n","\n","from glob import glob\n","from tqdm.auto import tqdm\n","import multiprocessing\n","\n","sequences_dir = \"/content/drive/MyDrive/Colabfold/sequences\"\n","results_dir = \"/content/drive/MyDrive/Colabfold/results\"\n","\n","def process_prediction(fasta_file, results_dir):\n","  print(f\"COMPUTING RESULTS FOR {os.path.basename(fasta_file)}\")\n","  config = get_config(fasta_file, results_dir=results_dir)\n","  run_job(config)\n","\n","workers_limit = 2  # Número máximo de trabajos en paralelo\n","with multiprocessing.Pool(processes=workers_limit) as pool:\n","    pool.starmap(process_prediction, [(fasta_file, results_dir) for fasta_file in glob(sequences_dir+\"/*.fasta\")])\n","\n"]}],"metadata":{"accelerator":"GPU","colab":{"provenance":[{"file_id":"1xtkw0Jafp0hBZRMgkTEKRIlUyR_2ug_I","timestamp":1694659831814},{"file_id":"17VU5NERUYSWStGrpBrItsZ74V3PPLwIt","timestamp":1693797915379},{"file_id":"https://github.com/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb","timestamp":1693262016788}],"machine_shape":"hm","gpuType":"V100"},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.10"}},"nbformat":4,"nbformat_minor":0}